File: tools/monpack.py
#!/usr/bin/env python3
"""怪物素材包的共用部分：對槽號、寫檔。

Mega Drive 與 Amiga 都要做同一件事 —— 把整張怪物圖烘成索引色 PNG，
再配上 DOS `MONSTERS.16` 的槽號。**槽號不能照檔案順序推**：

  - Mega Drive 的 nametable 在 ROM 裡的順序與 DOS 槽號是一個**排列**。
  - Amiga 的 `.anm` 接近恆等，但 DOS 有空槽而 Amiga 沒有，從第 42 槽起
    整批位移一格、第 73 槽起位移兩格。

兩邊都用同一個判準：拿**剪影**（哪些像素非透空）逐張比對 DOS 的基準圖，
再做貪婪一對一指派。一對一是重點 —— 只取「每個槽的最高分」會讓兩個槽
搶到同一張圖，而那種錯誤在總覽圖上看不出來。
"""

import json
import os

TRANSPARENT = 0


def dos_masks(data_dir: str):
    """回傳 {DOS 槽號: 剪影布林陣列}，尺寸統一補到 (86, 84)。"""
    import numpy as np

    import mm216

    blob = open(os.path.join(data_dir, "MONSTERS.16"), "rb").read()
    slots = mm216.monster_index(blob)
    out = {}
    for s in range(len(slots)):
        if not slots[s]:
            continue
        fr, _ = mm216.parse_monster(blob, s)
        _, _, w, h, px = fr[0]
        m = np.frombuffer(px, dtype=np.uint8).reshape(h, w) != mm216.MON_TRANSPARENT
        pad = np.zeros((86, 84), bool)
        pad[:min(h, 86), :min(w, 84)] = m[:86, :84]
        out[s] = pad
    return out


def assign(pic_masks, dos):
    """貪婪一對一指派。pic_masks 是 {圖索引: 剪影}，回傳 {圖索引: (槽號, 分數)}。

    **分數相同時取圖索引小的。** Mega Drive 的 ROM 裡有兩隻怪各存了兩份，
    剪影完全相同 —— 分數一路平手到小數點後十位，靠排序的自然順序決定
    等於讓「哪一份贏」隨實作細節浮動。先出現的那一份自帶調色盤
    （後面那一份的 tile 池沒有區塊頭，也就沒有調色盤，只能沿用上一張的），
    所以取小的那一份顏色才是對的。
    """
    import numpy as np

    scores = []
    for s, dm in dos.items():
        for i, pm in pic_masks.items():
            pad = np.zeros((86, 84), bool)
            h, w = pm.shape
            pad[:min(h, 86), :min(w, 84)] = pm[:86, :84]
            scores.append((float((dm == pad).mean()), s, i))
    # 分數大的優先；平手時圖索引小的優先（槽號只是讓排序穩定）。
    scores.sort(key=lambda t: (-t[0], t[2], t[1]))
    out, used_s, used_i = {}, set(), set()
    for sc, s, i in scores:
        if s in used_s or i in used_i:
            continue
        out[i] = (s, round(sc, 4))
        used_s.add(s)
        used_i.add(i)
    return out


def write(outdir: str, source: str, w: int, h: int, pics, slot_of) -> None:
    """把每張圖的每個影格存成索引色 PNG，外加 `set.json`。

    pics 是 [(圖索引, 調色盤, [影格的索引陣列, …])] 或
    [(圖索引, 調色盤, [影格…], [[影格號, 停留次數], …])] —— 第四項是
    平台自己的播放清單（Amiga 的 `.anm` 有動畫表，Mega Drive 沒有）。
    沒有第四項時 `set.json` 不寫 `anim`，讀取端就等速輪播。

    **一定要存索引色**（不是 RGBA）：Go 那邊 `png.Decode` 直接拿到
    `*image.Paletted`，正是 `render.Screen.BlitHiKey` 吃的型別。存 RGBA
    還要在 Go 裡反推調色盤，而反推會把剛好同色的兩個索引併掉。
    """
    from PIL import Image

    os.makedirs(outdir, exist_ok=True)
    entries = []
    for rec in pics:
        i, pal, frames = rec[0], rec[1], rec[2]
        play = rec[3] if len(rec) > 3 else None
        names = []
        for f, arr in enumerate(frames):
            im = Image.new("P", (w, h))
            im.putpalette([c for rgb in pal for c in rgb])
            im.putdata(arr.flatten().tolist())
            name = "pic%02d_f%02d.png" % (i, f)
            im.save(os.path.join(outdir, name), transparency=TRANSPARENT)
            names.append(name)
        s, sc = slot_of.get(i, (-1, 0.0))
        ent = {"pic": i, "slot": s, "match": sc, "frames": names}
        if play:
            ent["anim"] = play
        entries.append(ent)
    meta = {"source": source, "width": w, "height": h,
            "clear": TRANSPARENT, "pictures": entries}
    with open(os.path.join(outdir, "set.json"), "w", encoding="utf-8") as fh:
        json.dump(meta, fh, ensure_ascii=False, indent=1)
        fh.write("\n")
    n = sum(len(e["frames"]) for e in entries)
    mapped = sum(1 for e in entries if e["slot"] >= 0)
    print(f"{len(entries)} 張圖（對到槽號 {mapped}）、{n} 個影格 → {outdir}")

File: internal/assets/monpack/monpack.go
// Package monpack 讀烘好的怪物素材包（Mega Drive、Amiga）。
//
// 素材包由 `tools/mdgfx.py --export` 與 `tools/amiga32.py --export-monsters`
// 產生，格式相同：每張圖的每個影格一個索引色 PNG（索引 0 透空），
// 外加一份 `set.json`。尺寸由 `set.json` 宣告 —— Mega Drive 是 88×88、
// Amiga 是 84×86，不寫死。
//
// **為什麼是烘好的檔案而不是執行時解 ROM**：一張圖在 ROM 裡是「去重過的
// tile 池 ＋ 一張 nametable」，而 nametable 的格子順序是**硬體 sprite 的
// 順序**（九個 sprite，寬 4/4/3、高 4/4/3，sprite 內部直向排），不是
// row-major。那個版面是從實機的 sprite 屬性表影子（work RAM `0xFFD2A8`）
// 讀出來的 —— 既然已經定案，重建一次就好。推導見
// `docs/research/02-other-platforms.md`。
//
// 槽號用 DOS `MONSTERS.16` 的剪影對出來的：ROM 裡的順序與 DOS 的槽號
// **是一個排列不是恆等**，所以 `set.json` 的 `slot` 才是能用的鍵，
// PNG 的檔名編號不是。
package monpack

import (
	"encoding/json"
	"fmt"
	"image"
	"image/png"
	"os"
	"path/filepath"
)

// TransparentIndex 是透空的調色盤索引。ROM 裡每張圖的第 0 格就是背景色，
// 烘出來的 PNG 也把 0 標成 transparency。
const TransparentIndex = 0

// Set 是一整包，以 DOS 槽號為鍵。
type Set struct {
	// Source 是來源平台，給訊息用。
	Source string
	// W、H 是一張圖的原版像素尺寸。
	W, H int
	// Pics[槽號] 是那一槽的全部影格，至少一張。
	Pics map[int][]*image.Paletted
	// Extra 是對不到 DOS 槽的圖（片頭的旋轉地球、書、城堡那些）。
	// 留著是為了「解出來的東西不丟掉」，遊戲用不到。
	Extra [][]*image.Paletted
}

// Slots 回傳有圖的槽號個數。
func (s *Set) Slots() int { return len(s.Pics) }

type manifest struct {
	Source   string `json:"source"`
	Width    int    `json:"width"`
	Height   int    `json:"height"`
	Pictures []struct {
		Pic    int      `json:"pic"`
		Slot   int      `json:"slot"`
		Match  float64  `json:"match"`
		Frames []string `json:"frames"`
	} `json:"pictures"`
}

// Load 讀一個素材包目錄。目錄不存在時回傳錯誤 —— 呼叫端把它當成
// 「這個平台沒有怪物素材」處理，不是失敗。
func Load(dir string) (*Set, error) {
	b, err := os.ReadFile(filepath.Join(dir, "set.json"))
	if err != nil {
		return nil, err
	}
	var mf manifest
	if err := json.Unmarshal(b, &mf); err != nil {
		return nil, fmt.Errorf("set.json 解不開：%w", err)
	}
	if mf.Width <= 0 || mf.Height <= 0 {
		return nil, fmt.Errorf("set.json 沒宣告尺寸")
	}
	out := &Set{Source: mf.Source, W: mf.Width, H: mf.Height,
		Pics: map[int][]*image.Paletted{}}
	for _, p := range mf.Pictures {
		frames := make([]*image.Paletted, 0, len(p.Frames))
		for _, name := range p.Frames {
			im, err := loadPaletted(filepath.Join(dir, name), mf.Width, mf.Height)
			if err != nil {
				return nil, err
			}
			frames = append(frames, im)
		}
		if len(frames) == 0 {
			continue
		}
		if p.Slot < 0 {
			out.Extra = append(out.Extra, frames)
			continue
		}
		out.Pics[p.Slot] = frames
	}
	if len(out.Pics) == 0 {
		return nil, fmt.Errorf("%s 裡沒有對到 DOS 槽號的圖", dir)
	}
	return out, nil
}

// loadPaletted 讀一張索引色 PNG。
//
// **一定要是索引色**。存成 RGBA 也畫得出來，但畫面上的東西會少一層資訊：
// 兩個剛好同色的調色盤索引會被併成一個，而 `render` 的高解析貼圖走的是
// `*image.Paletted`。所以這裡直接把型別當契約檢查。
func loadPaletted(path string, w, h int) (*image.Paletted, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	im, err := png.Decode(f)
	if err != nil {
		return nil, fmt.Errorf("%s：%w", filepath.Base(path), err)
	}
	p, ok := im.(*image.Paletted)
	if !ok {
		return nil, fmt.Errorf("%s 不是索引色 PNG（是 %T）", filepath.Base(path), im)
	}
	if b := p.Bounds(); b.Dx() != w || b.Dy() != h {
		return nil, fmt.Errorf("%s 是 %d×%d", filepath.Base(path), b.Dx(), b.Dy())
	}
	return p, nil
}
